File: Problems/python/boj_연결요소의개수.py
#https://www.acmicpc.net/problem/11724
#DFS/BFS

from collections import deque
import sys
input = sys.stdin.readline #이거로 받아야 bfs에서 시간초과 안남...
sys.setrecursionlimit(10**7) #파이썬은 재귀제한이 걸려있어서 이걸로 조절해줘야한다.

n, m = map(int, input().split())
graph = [[] for _ in range(n+1)]

for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

visited = [False]*(n+1)

def dfs(graph, s, visited):
    visited[s] = True
    for i in graph[s]:
        if not visited[i]:
            dfs(graph, i, visited)

def bfs(graph, s, visited):
    q = deque([s])
    visited[s] = True
    while q:
        now = q.popleft()
        for i in graph[now]:
            if not visited[i]:
                visited[i] = True
                q.append(i)
        
result = 0
for i in range(1, n+1):
    if not visited[i]:
        #dfs(graph, i, visited)
        bfs(graph, i, visited)
        result += 1
    if len(set(visited[1:]))==1:
        break
        
print(result)

File: Problems/cpp/boj_연결요소의개수.cpp
//https://www.acmicpc.net/problem/11724
//DFS, BFS

#include <iostream>
#include <vector>
#include <deque>

using namespace std;

int n, m;
vector<vector<int>> graph;
vector<bool> visit;

void bfs(int start){
    visit[start] = true;
    deque<int> dq;
    dq.push_back(start);

    while(!dq.empty()){
        int now = dq.front();
        dq.pop_front();

        for(int i=0; i<graph[now].size(); i++){
            if(!visit[graph[now][i]]){
                visit[graph[now][i]] = true;
                dq.push_back(graph[now][i]);
            }
        }
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    int cnt = 0;

    cin >> n >> m;
    graph.resize(n+1);
    visit.resize(n+1);

    int a, b;
    for (int i=0; i<m; i++){
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    for(int i=1; i<n+1; i++){
        if(!visit[i]){
            bfs(i);
            cnt++;
        }
    }
    
    cout << cnt;
}